api/bag_layer.cpp

```diff
@@ -0,0 +1,270 @@
+
+\return
+    The size of the specified data type.
+*/
+uint8_t Layer::getElementSize(DataType type)
+{
+    switch (type)
+    {
+    case DT_FLOAT32:
+        return sizeof(float);
+    case DT_UINT32:
+        return sizeof(uint32_t);
+    case DT_UINT8:
+        return sizeof(uint8_t);
+    case DT_UINT16:
+        return sizeof(uint16_t);
+    case DT_UINT64:
+        return sizeof(uint64_t);
+    case DT_BOOLEAN:
+        return sizeof(bool);
+    case DT_STRING:
+        return sizeof(char*);
+    case DT_COMPOUND:  //[[fallthrough]]
+    case DT_UNKNOWN_DATA_TYPE:  //[[fallthrough]]
+    default:
+        throw UnsupportedElementSize{};
+    }
+}
+
+//! Retrieve the HDF5 path of the specified layer.
+/*!
+\param layerType
+    The type of layer.
+\param groupType
+    The type of group.
+    NODE or ELEVATION.
+
+\return
+    The HDF5 path to the specified layer and group types.
+*/
+std::string Layer::getInternalPath(
+    LayerType layerType,
+    GroupType groupType)
+{
+    if (groupType == NODE)
+        return NODE_GROUP_PATH;
+
+    if (groupType == ELEVATION)
+        return ELEVATION_SOLUTION_GROUP_PATH;
+
+    if (groupType != UNKNOWN_GROUP_TYPE)
+        throw UnsupportedGroupType{};
+
+    switch (layerType)
+    {
+    case Elevation:
+        return ELEVATION_PATH;
+    case Uncertainty:
+        return UNCERTAINTY_PATH;
+    case Hypothesis_Strength:
+        return HYPOTHESIS_STRENGTH_PATH;
+    case Num_Hypotheses:
+        return NUM_HYPOTHESES_PATH;
+    case Shoal_Elevation:
+        return SHOAL_ELEVATION_PATH;
+    case Std_Dev:
+        return STANDARD_DEV_PATH;
+    case Num_Soundings:
+        return NUM_SOUNDINGS_PATH;
+    case Average_Elevation:
+        return AVERAGE_PATH;
+    case Nominal_Elevation:
+        return NOMINAL_ELEVATION_PATH;
+    case Surface_Correction:
+        return VERT_DATUM_CORR_PATH;
+    case Compound:  // [[fallthrough]];
+    case UNKNOWN_LAYER_TYPE:  // [[fallthrough]];
+    default:
+        throw UnsupportedLayerType{};
+    }
+}
+
+//! Read a section of data from this layer.
+/*!
+    Read data from this layer starting at rowStart, columnStart, and continue
+    until rowEnd, columnEnd (inclusive).
+
+\param rowStart
+    The starting row.
+\param columnStart
+    The starting column.
+\param rowEnd
+    The ending row (inclusive).
+\param columnEnd
+    The ending column (inclusive).
+
+\return
+    The section of data specified by the rows and columns.
+*/
+UInt8Array Layer::read(
+    uint32_t rowStart,
+    uint32_t columnStart,
+    uint32_t rowEnd,
+    uint32_t columnEnd) const
+{
+    if (rowStart > rowEnd || columnStart > columnEnd)
+        throw InvalidReadSize{};
+
+    if (m_pBagDataset.expired())
+        throw DatasetNotFound{};
+
+    const auto pDataset = m_pBagDataset.lock();
+    uint32_t numRows = 0, numColumns = 0;
+    std::tie(numRows, numColumns) = pDataset->getDescriptor().getDims();
+
+    if (columnEnd >= numColumns || rowEnd >= numRows)
```

Comment:
This causes VRRefinments layer to not work since it doesn't have the same size as the dataset.

Reply:
This will be handled post merge. See: https://github.com/OpenNavigationSurface/BAG/issues/63